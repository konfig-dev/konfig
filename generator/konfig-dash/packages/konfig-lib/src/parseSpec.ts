import * as fs from 'fs-extra'
import * as path from 'path'
import * as os from 'os'
import * as yaml from 'js-yaml'

import SwaggerParser from '@apidevtools/swagger-parser'
import { OpenAPIV3, OpenAPIV3_1 } from 'openapi-types'
import { convertSwaggerToOAS3 } from './convertSwaggerToOAS3'

export type SchemaObject = OpenAPIV3.SchemaObject | OpenAPIV3_1.SchemaObject

export type PathsObject = OpenAPIV3.PathsObject | OpenAPIV3_1.PathsObject

export type RequestBodyObject =
  | OpenAPIV3.RequestBodyObject
  | OpenAPIV3_1.RequestBodyObject

export type MediaTypeObject =
  | OpenAPIV3.MediaTypeObject
  | OpenAPIV3_1.MediaTypeObject

export type ParameterObject =
  | OpenAPIV3.ParameterObject
  | OpenAPIV3_1.ParameterObject

export type OpenAPIV3_XDocument = OpenAPIV3.Document | OpenAPIV3_1.Document

export type ComponentsObject = NonNullable<OpenAPIV3_XDocument['components']>

export type ResponsesObject =
  | OpenAPIV3.ResponsesObject
  | OpenAPIV3_1.ResponsesObject

export type ResponseObject =
  | OpenAPIV3.ResponseObject
  | OpenAPIV3_1.ResponseObject

export type Spec = {
  spec: OpenAPIV3_XDocument
  $ref: SwaggerParser.$Refs
  specDereferenced: OpenAPIV3_XDocument | null
}

export const parseSpec = async (rawSpecString: string): Promise<Spec> => {
  const tmpDir = fs.mkdtempSync(path.join(os.tmpdir(), 'konfig-parseSpec'))
  const specFilePath = `${tmpDir}/spec`

  // see if rawSpecString is yaml and convert it to JSON before writing to specFilePath
  // we do this because yaml parsing actually supports dates which is problematic when
  let specJson: string | null = null
  try {
    specJson = JSON.stringify(JSON.parse(rawSpecString))
  } catch (e) {
    specJson = JSON.stringify(yaml.load(rawSpecString))
  }
  if (specJson === null) throw Error('Could not parse spec as JSON or YAML')
  fs.writeFileSync(specFilePath, specJson)

  // We specifically do not want to fully SwaggerParser#dereference the spec so
  // that generated SDKs generates and uses the provided refs & schemas (models)
  // SwaggerParser#parse is exactly what we want to call.
  const parser = new SwaggerParser()
  let spec = await parser.parse(specFilePath)

  // Handle Swagger / OAS2 by converting to OAS3
  if (!('openapi' in spec)) {
    spec = await convertSwaggerToOAS3({ spec })
    spec = await parser.parse(spec)
    if (!('openapi' in spec))
      throw Error('Could not successfuly convert to OAS version 3')
  }

  try {
    // its important to make directly call static resolve so that parser state
    // isnt shared with dereferenced spec
    let specDereferenced = await SwaggerParser.dereference(specFilePath, {
      dereference: { circular: 'ignore' },
    })
    if (!('openapi' in specDereferenced))
      specDereferenced = await convertSwaggerToOAS3({ spec: specDereferenced })
    return {
      spec,
      $ref: parser.$refs,
      specDereferenced,
    }
  } catch (e) {
    // Let pass
  }
  fs.removeSync(tmpDir)
  return {
    spec,
    $ref: parser.$refs,
    specDereferenced: null,
  }
}
