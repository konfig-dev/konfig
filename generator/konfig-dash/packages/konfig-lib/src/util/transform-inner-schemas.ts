import { Spec } from '../parseSpec'
import { resolveRef } from '../resolveRef'
import camelcase from './camelcase'
import {
  ArraySchemaObject,
  NonArraySchemaObject,
} from './generate-schema-object-from-json'

/**
 * Iteratively transforms all inner object or array type schemas into their own schema
 * in an OpenAPI Specification.
 * @param openApiDoc - The OpenAPI document to transform.
 * @returns The transformed OpenAPI document.
 */
export function transformInnerSchemas({ spec }: { spec: Spec }) {
  while (complexInnerSchemaDetected({ spec })) {
    extractComplexInnerSchema({ spec })
  }
}

function extractComplexInnerSchema({ spec }: { spec: Spec }) {
  visitComplexInnerSchema({
    spec,
    visit: ({ schemaName, schema, innerSchema, property }) => {
      if (spec.spec.components === undefined) return
      if (spec.spec.components.schemas === undefined) return
      const newSchemaName = generateNewSchemaName({ schemaName, property })
      const newSchemaRef = {
        $ref: `#/components/schemas/${newSchemaName}`,
      }
      // create new schema on #/components/schemas
      spec.spec.components.schemas[newSchemaName] = innerSchema
      // replace inner schema with $ref
      if (property) {
        if (schema.properties === undefined) return
        schema.properties[property] = newSchemaRef
      } else if (property === undefined) {
        schema.items = newSchemaRef
      }
    },
  })
}

function generateNewSchemaName({
  schemaName,
  property,
}: {
  schemaName: string
  property?: string
}) {
  if (property !== undefined)
    return `${schemaName}${camelcase(property, { pascalCase: true })}`
  return `${schemaName}Item`
}

function complexInnerSchemaDetected({ spec }: { spec: Spec }): boolean {
  return visitComplexInnerSchema({ spec, visit: () => true })
}

function visitComplexInnerSchema({
  spec: { spec, $ref },
  visit,
}: {
  spec: Spec
  visit: (
    args:
      | {
          innerSchema: NonArraySchemaObject | ArraySchemaObject
          property: string
          schema: NonArraySchemaObject
          schemaName: string
        }
      | {
          innerSchema: NonArraySchemaObject | ArraySchemaObject
          schema: ArraySchemaObject
          schemaName: string
          property: undefined
        }
  ) => boolean | void
}): boolean {
  if (spec.components === undefined) return false
  if (spec.components.schemas === undefined) return false
  for (const schemaName in spec.components.schemas) {
    const schema = resolveRef({
      refOrObject: spec.components.schemas[schemaName],
      $ref,
    })
    if (schema.type !== 'array' && schema.type !== 'object') continue
    if (schema.type === 'array') {
      const refOrObject = schema.items
      if ('$ref' in refOrObject) continue
      const innerSchema = resolveRef({
        refOrObject,
        $ref,
      })
      if (innerSchema.type === 'array' || innerSchema.type === 'object') {
        const visited = visit({
          innerSchema,
          schema,
          property: undefined,
          schemaName,
        })
        if (visited) return visited
      }
    }
    if (schema.type === 'object') {
      for (const property in schema.properties) {
        const refOrObject = schema.properties[property]
        if ('$ref' in refOrObject) continue
        const innerSchema = resolveRef({
          refOrObject,
          $ref,
        })
        if (innerSchema.type === 'array' || innerSchema.type === 'object') {
          const visited = visit({
            innerSchema,
            schema,
            property,
            schemaName,
          })
          if (visited) return visited
        }
      }
    }
  }
  return false
}
