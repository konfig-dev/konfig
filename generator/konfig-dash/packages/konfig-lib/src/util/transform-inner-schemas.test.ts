import { OpenAPIV3 } from 'openapi-types'
import { transformInnerSchemas } from './transform-inner-schemas'
import { parseSpec } from '../parseSpec'

describe('transformInnerSchemas', () => {
  it('transforms inner schemas in OpenAPI document', async () => {
    const originalOpenApiDoc: OpenAPIV3.Document = {
      openapi: '3.0.0',
      info: {
        title: 'My API',
        version: '1.0.0',
      },
      components: {
        schemas: {
          User: {
            type: 'object',
            properties: {
              name: {
                type: 'string',
              },
              address: {
                type: 'object',
                properties: {
                  street: {
                    type: 'string',
                  },
                  city: {
                    type: 'string',
                  },
                },
              },
              tags: {
                type: 'array',
                items: {
                  type: 'object',
                  properties: {
                    name: {
                      type: 'string',
                    },
                  },
                },
              },
              friends: {
                type: 'array',
                items: {
                  $ref: "#/components/schemas/User"
                }
              }
            },
          },
          OtherSchema: {
            type: 'object',
            properties: {
              prop1: {
                type: 'string',
              },
            },
          },
        },
        securitySchemes: {
          apiKey: {
            type: 'apiKey',
            name: 'X-API-Key',
            in: 'header',
          },
        },
      },
      paths: {
        '/users': {
          get: {
            operationId: 'getUsers',
            responses: {
              '200': {
                description: 'OK',
                content: {
                  'application/json': {
                    schema: {
                      $ref: '#/components/schemas/User',
                    },
                  },
                },
              },
            },
          },
          post: {
            operationId: 'createUser',
            security: [{ apiKey: [] }],
            requestBody: {
              content: {
                'application/json': {
                  schema: {
                    $ref: '#/components/schemas/User',
                  },
                },
              },
            },
            responses: {
              '200': {
                description: 'OK',
                content: {
                  'application/json': {
                    schema: {
                      $ref: '#/components/schemas/User',
                    },
                  },
                },
              },
            },
          },
        },
      },
    }

    const spec = await parseSpec(JSON.stringify(originalOpenApiDoc))
    transformInnerSchemas({ spec })
    expect(spec.spec).toMatchSnapshot()
  })
})
