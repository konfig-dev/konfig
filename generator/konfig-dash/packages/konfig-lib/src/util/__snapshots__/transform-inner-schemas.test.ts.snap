// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transformInnerSchemas transforms inner schemas in OpenAPI document 1`] = `
{
  "components": {
    "schemas": {
      "OtherSchema": {
        "properties": {
          "prop1": {
            "type": "string",
          },
        },
        "type": "object",
      },
      "User": {
        "properties": {
          "address": {
            "$ref": "#/components/schemas/UserAddress",
          },
          "friends": {
            "items": {
              "$ref": "#/components/schemas/User",
            },
            "type": "array",
          },
          "name": {
            "type": "string",
          },
          "tags": {
            "$ref": "#/components/schemas/UserTags",
          },
        },
        "type": "object",
      },
      "UserAddress": {
        "properties": {
          "city": {
            "type": "string",
          },
          "street": {
            "type": "string",
          },
        },
        "type": "object",
      },
      "UserTags": {
        "items": {
          "$ref": "#/components/schemas/UserTagsItem",
        },
        "type": "array",
      },
      "UserTagsItem": {
        "properties": {
          "name": {
            "type": "string",
          },
        },
        "type": "object",
      },
    },
    "securitySchemes": {
      "apiKey": {
        "in": "header",
        "name": "X-API-Key",
        "type": "apiKey",
      },
    },
  },
  "info": {
    "title": "My API",
    "version": "1.0.0",
  },
  "openapi": "3.0.0",
  "paths": {
    "/users": {
      "get": {
        "operationId": "getUsers",
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/User",
                },
              },
            },
            "description": "OK",
          },
        },
      },
      "post": {
        "operationId": "createUser",
        "requestBody": {
          "content": {
            "application/json": {
              "schema": {
                "$ref": "#/components/schemas/User",
              },
            },
          },
        },
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/User",
                },
              },
            },
            "description": "OK",
          },
        },
        "security": [
          {
            "apiKey": [],
          },
        ],
      },
    },
  },
}
`;
